Sources/Swift/Integrations/SessionReplay/SentryTouchTracker.swift

```diff
@@ -0,0 +1,149 @@
+            
+            switch touch.phase {
+            case .began:
+                info.start = newEvent
+            case .ended, .cancelled:
+                info.end = newEvent
+            case .moved:
+                if let last = info.movements.last, touchesDelta(last.point, position) < 10 { continue }
+                info.movements.append(newEvent)
+                debounceEvents(in: info)
+            default:
+                continue
+            }
+            
+            trackedTouches[touch] = info
+        }
+    }
+    
+    private func touchesDelta(_ lastTouch: CGPoint, _ newTouch: CGPoint) -> CGFloat {
+        let dx = newTouch.x - lastTouch.x
+        let dy = newTouch.y - lastTouch.y
+        return sqrt(dx * dx + dy * dy)
+    }
+    
+    private func debounceEvents(in touchInfo: TouchInfo) {
+        guard touchInfo.movements.count >= 3 else { return }
+        let subset = touchInfo.movements.suffix(3)
+        if subset[subset.startIndex + 2].timestamp - subset[subset.startIndex + 1].timestamp > 0.5 {
+            //Dont debounce if the touch has a 1 second difference to show this pause in the replay.
+            return
+        }
+        if arePointsCollinearSameDirection(subset[subset.startIndex].point, subset[subset.startIndex + 1].point, subset[subset.startIndex + 2].point) {
+            touchInfo.movements.remove(at: touchInfo.movements.count - 2)
+        }
+    }
+    
+    private func arePointsCollinearSameDirection(_ a: CGPoint, _ b: CGPoint, _ c: CGPoint, tolerance: CGFloat = 10) -> Bool {
+        let crossProduct = (b.y - a.y) * (c.x - b.x) - (c.y - b.y) * (b.x - a.x)
+        
+        if abs(crossProduct) >= tolerance {
+            return false //Not collinear, return early
+        }
+              
+        let directionAB = CGPoint(x: b.x - a.x, y: b.y - a.y)
+        let directionBC = CGPoint(x: c.x - b.x, y: c.y - b.y)
+        let lengthAB = sqrt(pow(directionAB.x, 2) + pow(directionAB.y, 2))
+        let lengthBC = sqrt(pow(directionBC.x, 2) + pow(directionBC.y, 2))
+        let normalizedAB = CGPoint(x: directionAB.x / lengthAB, y: directionAB.y / lengthAB)
+        let normalizedBC = CGPoint(x: directionBC.x / lengthBC, y: directionBC.y / lengthBC)
+        let dotProduct = normalizedAB.x * normalizedBC.x + normalizedAB.y * normalizedBC.y
+        
+        // Directions are considered the same if the dot product is close to 1 (with some tolerance)
+        return abs(dotProduct - 1) < 0.001
+    }
+  
+    func flushFinishedEvents() {
+        trackedTouches = trackedTouches.filter { $0.value.end == nil }
+    }
+    
+    func replayEvents(from: Date, until: Date) -> [SentryRRWebEvent] {
+        let uptime = dateProvider.systemUptime()
+        let now = dateProvider.date()
+        let startTimeInterval = uptime - now.timeIntervalSince(from)
+        let endTimeInterval = uptime - now.timeIntervalSince(until)
```

Comment:
`m`: It took me quite a while to figure out why we need sytemUptime here. It's because of `UIEvent.timestamp`

> This property contains the number of seconds that have elapsed since system startup. For a description of this time value, see the description of the [systemUptime](https://developer.apple.com/documentation/foundation/processinfo/1414553-systemuptime) method of the [ProcessInfo](https://developer.apple.com/documentation/foundation/processinfo) class.
> from Apple Docs: https://developer.apple.com/documentation/uikit/uievent/1613790-timestamp

Instead of converting the `timestamp` around multiple times in this method with `now.addingTimeInterval(infoStart.timestamp - uptime)`, can't we just directly store the timestamp of type Date on the `TouchEvent` in `trackTouchFrom`?

Reply:
We could, which technically would result in the same number of conversions in the long run. 
I chose not to because, with this approach, we're safeguarded against time drift—for example, if the timezone changes in the middle of the replay segment

Reply:
> I chose not to because, with this approach, we're safeguarded against time drift—for example, if the timezone changes in the middle of the replay segment

That makes sense. It would be great to add this info as a comment and also a short hint on why we have to use systemUptime at the beginning of this method would greatly improve the readability of this code.